src/uct/ib/rdmacm/rdmacm_cm_ep.c

```diff
@@ -344,52 +335,122 @@ static ucs_status_t uct_rdamcm_cm_ep_server_init(uct_rdmacm_cm_ep_t *cep,
                            uct_cm_ep_server_conn_notify_callback_t,
                            ucs_empty_function);
     if (status != UCS_OK) {
-        goto err;
+        goto err_reject;
     }
 
     cep->id->context = cep;
 
-    memset(&conn_param, 0, sizeof(conn_param));
-    conn_param.private_data = ucs_alloca(uct_rdmacm_cm_get_max_conn_priv() +
-                                         sizeof(uct_rdmacm_priv_data_hdr_t));
+    if (ucs_test_all_flags(params->field_mask,
+                           UCT_EP_PARAM_FIELD_PRIV_DATA |
+                           UCT_EP_PARAM_FIELD_PRIV_DATA_LENGTH)) {
+        priv_data        = params->private_data;
+        priv_data_length = params->private_data_length;
+    } else if (params->field_mask & UCT_EP_PARAM_FIELD_SOCKADDR_PACK_CB) {
+        status = uct_rdmacm_cm_ep_pack_cb(cep, pack_priv_data,
+                                          &pack_priv_data_length);
+        if (status != UCS_OK) {
+            goto err_reject;
+        }
+
+        priv_data        = &pack_priv_data;
+        priv_data_length = pack_priv_data_length;
+    } else {
+        priv_data        = NULL;
+        priv_data_length = 0;
+    }
 
-    status = uct_rdmacm_cm_ep_pack_cb(cep, &conn_param);
+    status = uct_rdmacm_cm_ep_send_priv_data(cep, priv_data, priv_data_length);
     if (status != UCS_OK) {
-        goto err_reject;
+        goto err;
+    }
+
+    return uct_rdmacm_cm_ack_event(event);
+err_reject:
+    uct_rdmacm_cm_reject(cep->id);
+err:
+    uct_rdmacm_cm_destroy_id(cep->id);
+    cep->id = NULL;
+    uct_rdmacm_cm_ack_event(event);
+    return status;
+}
+
+ucs_status_t
+uct_rdmacm_cm_ep_send_priv_data(uct_rdmacm_cm_ep_t *cep, const void *priv_data,
+                                size_t priv_data_length)
+{
+    struct rdma_conn_param conn_param = {0};
+    uct_rdmacm_priv_data_hdr_t *hdr;
+    uct_cm_remote_data_t remote_data;
+    char ep_str[UCT_RDMACM_EP_STRING_LEN];
+    ucs_status_t status;
+
+    if (priv_data_length > uct_rdmacm_cm_get_max_conn_priv()) {
+        status = UCS_ERR_EXCEEDS_LIMIT;
+        goto err;
     }
 
     status = uct_rdamcm_cm_ep_set_qp_num(&conn_param, cep);
     if (status != UCS_OK) {
-        goto err_reject;
+        goto err;
     }
 
-    ucs_trace("%s: rdma_accept on cm_id %p",
-              uct_rdmacm_cm_ep_str(cep, ep_str, UCT_RDMACM_EP_STRING_LEN),
-              event->id);
+    conn_param.private_data     = ucs_alloca(UCT_RDMACM_TCP_PRIV_DATA_LEN);
+    conn_param.private_data_len = sizeof(*hdr);
 
-    if (rdma_accept(event->id, &conn_param)) {
-        uct_cm_ep_peer_error(&cep->super, "rdma_accept(on id=%p) failed: %m",
-                             event->id);
-        uct_rdmacm_cm_ep_destroy_dummy_qp(cep);
-        status = UCS_ERR_IO_ERROR;
-        goto err;
+    hdr         = (uct_rdmacm_priv_data_hdr_t*)conn_param.private_data;
```

Comment:
init inline

Reply:
`conn_param.private_data` is just allocated above

Reply:
ok